src/main/java/io/tus/java/client/TusUploader.java

```diff
@@ -159,6 +160,47 @@ public int getRequestPayloadSize() {
         return requestPayloadSize;
     }
 
+    /**
+     * Upload a part of the file by reading a chunk from the InputStream and writing
+     * it to the HTTP request's body. If the number of available bytes is lower than the chunk's
+     * size, all available bytes will be uploaded and nothing more.
+     * The size of the read chunk can be obtained using {@link #getRequestPayloadSize()} and changed
+     * using {@link #setRequestPayloadSize(int)}.
+     * In order to obtain the new offset, use {@link #getOffset()} after this method returns.
+     *
+     * @return Number of bytes read and written.
+     * @throws IOException  Thrown if an exception occurs while reading from the source or writing
+     *                      to the HTTP request.
+     */
+    public int upload() throws IOException, ProtocolException {
```

Comment:
I don't think having separate `upload` and `uploadChunk` methods is a good API design. First off all, if you end up calling both methods on the same TusUploader instance, you will likely get silent problems since the connection's streaming mode is not changed.

Instead, I would prefer if the chunked streaming mode could be controlled using a flag in TusClient just like `enableRemoveFingerprintOnSuccess` etc.

Reply:
As a user, I was disturbed by the two sizes: one for the chunks and the other for the whole payload length. So it makes sense to avoid complicating the API.

As a newbie, I wish such things occurs automagically: an initial hand-shake (HEAD? OPTIONS?) will let the client discover if the server support chunked transfer. But this requires much more work.

I will rework my patch in order to control this behavior from TusClient. What about `enableChunkedTransferEncoding()` and `disableChunkedTransferEncoding()`?

Reply:
> As a user, I was disturbed by the two sizes: one for the chunks and the other for the whole payload length. 

I absolutely agree. There are plans to remove the chunk size setting in a new major release but there is no time plan at the moment.

> As a newbie, I wish such things occurs automagically: an initial hand-shake (HEAD? OPTIONS?) will let the client discover if the server support chunked transfer. But this requires much more work.

It would make things easier but also slows uploads down if a handshake is required. So preconfiguring is preferred from a performance point of view.

> I will rework my patch in order to control this behavior from TusClient. What about enableChunkedTransferEncoding() and disableChunkedTransferEncoding()?

Sounds good!